File: output/GPT3.5-Turbo/xCodeEval_4000/C++/Go/b1cedf95a649a34372e483ba529e37c8.go
package main

import (
	"bufio"
	"fmt"
	"os"
	"strconv"
	"strings"
)

func main() {
	scanner := bufio.NewScanner(os.Stdin)
	scanner.Split(bufio.ScanWords)

	scanner.Scan()
	N, _ := strconv.Atoi(scanner.Text())

	scanner.Scan()
	M, _ := strconv.Atoi(scanner.Text())

	K := make([]int, N)
	for i := 0; i < N; i++ {
		scanner.Scan()
		K[i], _ = strconv.Atoi(scanner.Text())
	}

	vv := make([][]int, 4*e5+1)
	for i := 0; i < M; i++ {
		scanner.Scan()
		D, _ := strconv.Atoi(scanner.Text())

		scanner.Scan()
		T, _ := strconv.Atoi(scanner.Text())

		vv[D-1] = append(vv[D-1], T-1)
	}

	ok := 2 * sum(K)
	ng := sum(K) - 1

	for ok-ng > 1 {
		k := (ok + ng) >> 1
		v := make([]int, k)
		b := make([]bool, N)

		for i := k - 1; i >= 0; i-- {
			for _, j := range vv[i] {
				if !b[j] {
					b[j] = true
					v[i] += K[j]
				}
			}
		}

		t := 0
		s := sum(K)

		for i := 0; i < k; i++ {
			t++
			m := min(t, v[i])
			t -= m
			s -= m
		}

		if t >= s*2 {
			ok = k
		} else {
			ng = k
		}
	}

	fmt.Println(ok)
}

func sum(a []int) int {
	s := 0
	for _, e := range a {
		s += e
	}
	return s
}

func min(a, b int) int {
	if a < b {
		return a
	}
	return b
}

File: output/GPT3.5-Turbo/xCodeEval_4000/C++/Python/b1cedf95a649a34372e483ba529e37c8.py
import sys
from typing import List

def main():
    N, M = map(int, input().split())
    K = list(map(int, input().split()))
    vv = [[] for _ in range(4 * 10**5 + 1)]
    for _ in range(M):
        D, T = map(int, input().split())
        vv[D - 1].append(T - 1)
    ok = 2 * sum(K)
    ng = sum(K) - 1
    while ok - ng > 1:
        k = (ok + ng) // 2
        v = [0] * k
        b = [False] * N
        for i in range(k - 1, -1, -1):
            for j in vv[i]:
                if not b[j]:
                    b[j] = True
                    v[i] += K[j]
        t = 0
        s = sum(K)
        for i in range(k):
            t += 1
            m = min(t, v[i])
            t -= m
            s -= m
        if t >= s * 2:
            ok = k
        else:
            ng = k
    print(ok)

if __name__ == "__main__":
    main()
